shell/src/main/java/alluxio/shell/command/CpCommand.java

```diff
@@ -0,0 +1,236 @@
+   *
+   * @param srcPath the source {@link AlluxioURI} (could be a file or a directory)
+   * @param dstPath the destination path in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copy(AlluxioURI srcPath, AlluxioURI dstPath) throws IOException {
+    URIStatus srcStatus;
+    try {
+      srcStatus = mFileSystem.getStatus(srcPath);
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    URIStatus dstStatus = null;
+    try {
+      dstStatus = mFileSystem.getStatus(dstPath);
+    } catch (FileDoesNotExistException e) {
+      // if the destination does not exist, it will be created
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    if (srcStatus.isFolder()) {
+      if (dstStatus != null && !dstStatus.isFolder()) {
+        throw new IOException(
+            ExceptionMessage.DESTINATION_FILE_CANNOT_EXIST_WITH_WILDCARD_SOURCE.getMessage());
+      }
+
+      if (dstStatus == null) {
+        try {
+          mFileSystem.createDirectory(dstPath);
+          System.out.println("Created directory: " + dstPath.getPath());
+        } catch (AlluxioException e) {
+          throw new IOException("Fail to create directory: " + dstPath.getPath());
+        }
+      }
+
+      List<URIStatus> statuses;
+      try {
+        statuses = mFileSystem.listStatus(srcPath);
+      } catch (AlluxioException e) {
+        throw new IOException(e.getMessage());
+      }
+
+      List<String> errorMessages = new ArrayList<>();
+      for (URIStatus status : statuses) {
+        try {
+          copy(new AlluxioURI(srcPath.getScheme(), srcPath.getAuthority(), status.getPath()),
+              new AlluxioURI(dstPath.getScheme(), dstPath.getAuthority(),
+                  PathUtils.concatPath(dstPath.getPath(), status.getName())));
+        } catch (IOException e) {
+          errorMessages.add(e.getMessage());
+        }
+      }
+
+      if (errorMessages.size() != 0) {
+        throw new IOException(Joiner.on('\n').join(errorMessages));
+      }
+    } else {
+      copyFile(srcPath, dstPath);
+    }
+  }
+
+  /**
+   * Copies a file in the Alluxio filesystem.
+   *
+   * @param srcPath the source {@link AlluxioURI} (has to be a file)
+   * @param dstPath the destination path in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copyFile(AlluxioURI srcPath, AlluxioURI dstPath) throws IOException {
+    URIStatus srcStatus;
+    try {
+      srcStatus = mFileSystem.getStatus(srcPath);
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    try (Closer closer = Closer.create()) {
+      OpenFileOptions openFileOptions = OpenFileOptions.defaults().setReadType(ReadType.NO_CACHE);
+      FileInStream is = closer.register(mFileSystem.openFile(srcPath, openFileOptions));
+      CreateFileOptions createFileOptions = CreateFileOptions.defaults();
+      if (srcStatus.isPersisted()) {
```

Comment:
When writing the cross-mount move job we decided that it makes more sense to respect the write type configuration over the existing state, so if we want to be consistent we should do the same here